matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/presence/service/task/SetPresenceTask.kt

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2020 The Matrix.org Foundation C.I.C.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.matrix.android.sdk.internal.session.presence.service.task
+
+import org.matrix.android.sdk.internal.network.GlobalErrorReceiver
+import org.matrix.android.sdk.internal.network.executeRequest
+import org.matrix.android.sdk.internal.session.presence.PresenceAPI
+import org.matrix.android.sdk.internal.session.presence.messages.PresenceEnum
+import org.matrix.android.sdk.internal.session.presence.messages.SetPresenceBody
+import org.matrix.android.sdk.internal.task.Task
+import javax.inject.Inject
+
+internal abstract class SetPresenceTask : Task<SetPresenceTask.Params, Any> {
+    data class Params(
+            val userId: String,
+            val presence: PresenceEnum,
+            val message: String?
+    )
+}
+
+internal class DefaultSetPresenceTask @Inject constructor(private val presenceAPI: PresenceAPI,
+                                                          private val globalErrorReceiver: GlobalErrorReceiver) : SetPresenceTask() {
+
+    override suspend fun execute(params: Params): Any {
+        return executeRequest(globalErrorReceiver) {
+            params.apply {
```

Comment:
weird way to write it, generally we just use `params.` which is I think a bit more readable.
As a general rule, we avoid using `apply` when it's not strictly necessary, because it may lead to hard to find bug: in the apply block you've got 2 different `this`

Reply:
Nice, let's stick to that convention then. Furthermore, when we use `apply` we can explicitly use `this@ClassName` to be more clear.

Reply:
Yes, I know, but it's quite ugly to me.

I want to avoid mistake which can happen for instance in this simple case:

```kotlin
class B {
    var myVar: Int = 0
}

class A {
    var b = B()
    var myVar: Int = 0

    fun test() {
        myVar = 4
        b.apply {
            myVar = 5
        }
    }
}
```

This code is compiling fine, without any warning, but developer can make mistake and/or have doubt about the value of `A.myVar` once the function `test` is called.

Reply:
Yes, I agree it might be confusing for someone so lets avoid it